### ISSUE TITLE
Refactor distributions

### ISSUE THREAD

**Issue by geraintpalmer** [CiwPython/Ciw#134]
Refactor the way in which distributions are done. (thanks @drvinceknight)

Class for each distributions (inheriting from an empty `Distribution` class). All they would need is a `.sample()` method, which may or may not take in a time `t` (for Time Dependent distributions).

This would make it much easier to implement own distributions, and new distributions such as Phase Type (#73).

(Big job).

**Comment by drvinceknight** [comment]
I would suggest this gets done in two steps (potentially two PRs):

1. Implement a distribution class and create classes for all distributions currently in the library;
2. Perform a distribution-octomy: replace the distributions in the library with these distributions classes.

Doing it in two seperate PRs should make reviewing simpler :)

**Comment by drvinceknight** [comment]
On #137, it became apparent that users might want to combine samples of distributions. This will be "straightforward" to do with distributions by implementing algebraic operations as methods on the class:

```
def __add__
def __mul__
```

So a user could do:

```
...
Service_distributions[ciw.uniform(...) + ciw.exponential(...)]]
...
```

as  a way to combine distributions.

**Comment by daffidwilde** [comment]
Following discussion with @geraintpalmer @drvinceknight, distributions will be instances of classes from a `distributions` sub-library. So users will do something like this:

```
>>> import ciw
>>> import ciw.distributions as dist

>>> N = ciw.create_network(
...     arrival_distributions=[dist.Uniform(3.4, 3.5),
...                            dist.Exponential(5)],
...     service_distributions=[dist.Exponential(10),
...                            dist.Deterministic(3)],
...     number_of_servers=[2, 2],
...     transition_matrix=[[0.1, 0.0],
...                        [0.5, 0.2]]
... )
```

**Comment by drvinceknight** [comment]
Information on operators: https://docs.python.org/2/library/operator.html

**Comment by daffidwilde** [comment]
> ...  users might want to combine samples of distributions. This will be "straightforward" to do with distributions by implementing algebraic operations as methods ...

This will be done nicely with instances of classes and the `operator` module:

```p
>>> import operator as op

>>> class Combined:
...     def __init__(self, thing, other, operator):
...     ...
...     def __repr__(self):
...         return f"{repr(self.thing)} {repr(self.operator)} {repr(self.other)}"
... 
...     def sample(self):
...         return self.operator(self.thing.sample(), self.other.sample())

>>> class BaseDistribution:
...     ...
...     def __add__(self, other):
...         return Combined(self, other, op.add)
...     def __mul__(self, other):
...         return Combined(self, other, op.mul)
```

**Comment by daffidwilde** [comment]
https://stackoverflow.com/questions/36539460/implementing-negation-by-overriding-neg-in-python

**Comment by geraintpalmer** [comment]
@drvinceknight @daffidwilde 

When defining a network and then a simulation, what 'owns' the instance of the distribution instance? the Network or the Simulation?

+ `ciw.Network` is designed to be how the system to be modelled is designed. Any amount of `ciw.Simulation` objects can be created by following the instructions contained in the `ciw.Network`.
+ `ciw.Simulation` is designed to be used once, as an engine for the simulation.

I think it makes more sense for the `Simulation` to have the distribution object. And to have different instances of the distribution object for every `Simulation` created. But the following code:

```
>>> N = ciw.create_network(
...     arrival_distributions=[dist.Uniform(3.4, 3.5),
...                            dist.Exponential(5)],
...     service_distributions=[dist.Exponential(10),
...                            dist.Deterministic(3)],
...     number_of_servers=[2, 2],
...     transition_matrix=[[0.1, 0.0],
...                        [0.5, 0.2]]
... )
```

implies that the `Network` object has the distribution object. And then the same object is passed to every `Simulation` object. This might cause problems, for example is attributes of the distribution object change over the run of a simulation, those *changed* attributed will be passed to the next trial.

What do you think @daffidwilde @drvinceknight ? Am I making sense?

**Comment by daffidwilde** [comment]
I think it makes more sense from the interface perspective to pass the instances in the `Network` object. To get around the issues you've raised, I can think of two solutions:

1. Creating a new instance of the distribution with the same parameters for each `Simulation` using `vars`:
```p
simulation_arrival_distributions = [
    dist.__class__(**vars(dist)) for dist in arrival_distributions
]
```

2. Writing a general `reset` method for distributions that is called at the instantiation of the `Simulation` object. This could be done by storing the original params at the instantiation of the distribution object(s):
```p
class BaseDistribution:

    def __init__(self):
        ...
        self.original_params = dict(vars(self))

    ...

    def reset(self):
        self.__dict__.update(self.original_params)
```

I think I prefer the latter as the instance of the class is in the same place rather than generating new, identical instances for each simulation.

**Comment by drvinceknight** [comment]
I agree from an API pov that it should be in the network.

I like @daffidwilde's `reset` method suggestion.

**Comment by geraintpalmer** [comment]
I'm not sure a 'resets' method would work:

+ means you can't parallelise, as all 'Simulations' objects will be
pointing to the same dist object
+ goes a little agains the rest of Ciw's api where the 'Simulations' object
stores the final state of the run after the simulation run. Allows starting
and stopping the simulation run manually.

My initial idea was a deepcopy, which seems similar to @daffwilde's first
suggestion, but his seems a little neater.

**Comment by drvinceknight** [comment]
> means you can't parallelise, as all 'Simulations' objects will be pointing to the same dist object

I'm not sure I follow/agree but I'm happy either way, as long as the api is as you wrote here: https://github.com/CiwPython/Ciw/issues/134#issuecomment-487630471

**Comment by geraintpalmer** [comment]
Implemented in commit dded0268b0500266bd726 with contribution from @daffidwilde
